### ISSUE TITLE
Rework activation / installation flow.

### ISSUE THREAD

**Issue by jdevalk** [Yoast/License-Manager#61]
## The problem
All  of our premium "plugins" are actually "add-ons". They will not function without their "parent" plugin, WordPress SEO, Google Analytics by Yoast, or something else. WordPress SEO Premium _seems_ like it's an exception, except that it's not: it's WordPress SEO with an extra folder of functionality. 

The installation process for these premium plugins is cumbersome at the very least. After the customer purchases one of these plugins, he/she gets an email with a zip file and a license key. He/she has to download the zip file, install it in their WordPress install and then enter their license key. 

We should take all this friction away.

## The solution
The "parent" plugin has an extensions page. On this extensions page, the customer should be able to see which extensions are available. If he wants to buy one, he clicks on a link and goes to our website to buy the extension. Once this is done, the customer gets a license key, straight on the thank you page.

The customer enters this license key in the field on the extensions page, and hits "activate" (or whatever we decide to put on that label). The plugin should at that point install the premium add-on. This add-on should be in a `yoast-addons` folder in the plugins folder so it doesn't get deleted when the plugin is updated. 

This is far from trivial, but seems like an incredible improvement to me. 

Please give me all your feedback, @tacoverdo @dannyvankooten @Rarst @omarreiss @andizer @Petervw @jrfnl.

**Comment by dannyvankooten** [comment]
Sounds like a huge improvement to me. 

First thing that comes to mind right now is that WordPress core completely wipes a plugin folder before extracting the .zip file of the update package. That would mean the `yoast-addons` folder would be wiped as well. 

Maybe a first iteration could be to stick with the regular way of managing plugins but have "free" download the initial package (after customer enters his key) and manage the updates for all packages?

**Comment by jdevalk** [comment]
That's why yoast-addons should be in the plugins folder, not in the folder
of the parent plugin.

**Comment by jdevalk** [comment]
But yeah that could be an easy first step!

**Comment by dannyvankooten** [comment]
Ah, of course. Didn't pick up on that because of this. :smile: 

> This add-on should be in a yoast-addons folder in the plugins folder

**Comment by Rarst** [comment]
> First thing that comes to mind right now is that WordPress core completely wipes a plugin folder

I think Joost meant pluginS folder, as in `wp-content/plugins`.

I am very wary of the writing things around and in "special" places, since that would need to go through Filesystem API, request credentials as necessary, complicate "manual" install...

I guess my main question — while make "special" addon instead of just "normal" plugin, providing the extra functionality? Just install it alongside the main plugin and have it work with it.

By the way, are "in-app" purchases in line with official WP repo rules in first place?

**Comment by uprise10** [comment]
Personally I think separate plugin directories as it is right now is just a fine setup, however the installation process could be easier.

I like the way [Gravity Perks](http://gravitywiz.com/) handles it. In fact these are separate plugins, but the installation and activation is done via the settings screen of this plugin.

**Comment by omarreiss** [comment]
There are different reasons why I think it would be a good idea to automate add-on installation and handle it in the extensions manager of mothership plugins like WordPress SEO and others:

- Major improvement of the user (onboarding) experience.
- We get to control which version of each add-on is installed with which version of the mothership plugin.
- Semantics

#### Install/Update UX

I think it's not hard to agree there's a lot of room for improvement when it comes to the UX for installing and updating premium add-ons. @jdevalk explains quite clearly what's suboptimal about that flow. When I compare that with the flow users go through when installing or updating a plugin from WordPress.org, I think we should really do better. 

#### Control

One of the ideas we had today with regard to the approach proposed here, was that we could make it impossible to update the add-ons separately from the mothership plugin. The mothership plugin could decide with which versions of each add-on it ships. This makes the plugins and add-ons we use so much more maintainable, since we don't have to worry anymore if that _new_ version of Local SEO will still be compatible with that _old_ version of WordPress SEO so many users are still on. As plugin authors, we gain control over those conditions (whereby we also regain some agility), and the user never has to worry about updating add-ons again. The promise towards the user simply becomes that he'll always get a plugin that is compatible with its add-ons and vice versa.

#### Semantics

So what is it you buy, when you buy an add-on? Is it a standalone plugin that integrates with the mothership plugin, or is it a tailored version of the mothership plugin, enhanced with those extra features you need on your site? I'd like to see add-ons as modules belonging to a __single__ plugin. The plugin also figures as a brand that covers a certain domain or problem space within WordPress. The plugin add-ons or modules enhance the plugin by covering niche aspects of that same domain. 

#### Technical challenges

Technically, I can imagine this could be very demanding, but if WP core can download a plugin from wordpress.org and write it to the plugins directory, then shouldn't this also mean that we could do a similar thing for add-ons at the same level of convenience for the user?

@Rarst Manual installs could be provided for by design if we wanted to. Seems like something that could be solved by turning all add-ons into composer packages?

**Comment by jdevalk** [comment]
@Rarst:

> By the way, are "in-app" purchases in line with official WP repo rules in first place?

We won't do in-app purchase. We'll link to the page on Yoast.com. EDD does exactly that, we already do this too. 

@uprise10:

> I like the way Gravity Perks handles it.

Didn't know that one, I'll have a look. 

In general: I very much dislike the whole separate plugin directory thing _because_ it uses the normal WordPress.org update process. It's "wrong", as long as WP doesn't properly support child plugins like it supports child themes. 

That being said, I like @dannyvankooten's idea of making a simple first iteration of this by just using the normal install process but doing it all for the user, instead of giving him a zip file. That'd already take 75% of the friction away.

So this would mean: input field in the main plugin, activation would lead not just to activation but would mean receiving the license manager would receive a download link from the license server, which the license manager would then use to install & activate the plugin.

**Comment by tacoverdo** [comment]
First of all, yes. I'd love to have a simpler UX. No doubt there.
So let me reflect on all of your proposals.

## Finding add-ons
Having a screen that lists all available/installed/activated add-ons (improvement of what we currently have) sounds like a very good idea. We might even move that page to 'General', so that users encounter it more often. 
From that 'showcase' page, we could link users to the correct sales pages on yoast.com or maybe even purchase right from the plugin (as far as I understood, this is not prohibited by the [dotorg rules](https://wordpress.org/plugins/about/guidelines/)).

## Different methods of installation
As @Rarst said, some installations/hosts have strange 'security measures' in place (or just downright incorrect setups). This sometimes causes users to enter FTP details for every plugin action they want to do. If we automate the install, we should be aware of this.
There are also quite a lot of users that do not use the WP admin to install plugins, but prefer (S)FTP. Or maybe even a plugin like WPRemote. We should make sure our premium plugins are also compatible with those.

## Control versions
From a developer point of view, I understand what @omarreiss is saying. However, this means I would (for example) have to install WordPress SEO 2.1 when all I want to do is update the Local SEO add-on I paid for. I don't think this is clear UX.

## My suggestion / summary
Make the add-ons easy to find and install (see above and suggestions by @jdevalk) as separate plugin. When they've purchased/installed a new add-on, take the user to an about-screen – much like we've implemented in the latest version of WordPress SEO – and let them fill in their license key there.

**Comment by Rarst** [comment]
> We won't do in-app purchase. We'll link to the page on Yoast.com. EDD does exactly that, we already do this too.

Yeah, not quite purchase, was typing/thinking late. Still I think there was some history against "plugins installing plugins" with official repository (Woo's mothership plugin maybe?), so I think that's something to consider.

Overall I am all for automation (very much so), but also for familiarity.

In my opinion "we installed additional plugin for you" is much easier on user (semantically and technically) than "we created special folder in your installation and put a totally new kind of a thing in there".

**Comment by jdevalk** [comment]
Re @tacoverdo's comments on controlling versions: I actually want to prohibit people from updating Local SEO and not WordPress SEO. That leads to headaches I simply don't want to solve.

**Comment by jdevalk** [comment]
So I've made #62 for a first iteration of this. Let's start with that.

**Comment by coenjacobs** [comment]
> From a developer point of view, I understand what @omarreiss is saying. However, this means I would (for example) have to install WordPress SEO 2.1 when all I want to do is update the Local SEO add-on I paid for. I don't think this is clear UX.

This all comes down to (unfortunately) WordPress' Achilles' heel: Dependency management. @yoast mentioned it already, WordPress doesn't allow child plugins. Requiring a certain version of plugin x in order to activate plugin y is definitely something that's doable, but there are more challenges there.

If you can solve this for WordPress based projects, you'll make a lot of people (including me) really happy. This is probably a different challenge than the one being discussed in this issue though.

**Comment by barrykooij** [comment]
@coenjacobs the problem with dependency management in WP atm is that core doesn't have it so you can't control other plugins dependencies. In the case where you control all dependencies (e.g. yoast main + yoast addons) this is definitely doable and very interesting for plugins with similar models.

**Comment by Rarst** [comment]
I've worked through the repository guidelines and transferring code is quite vocally prohibited in current state of them:

> 8. No sending executable code via third-party systems. Use of third-party systems is acceptable in a service-client type of model, but sending actual PHP or other executable code over the network is considered a security risk. Executing outside code like this is not allowed except for specific and very carefully considered cases (such as during upgrades, for example).

> https://wordpress.org/plugins/about/guidelines/

**Comment by coenjacobs** [comment]
That's probably the same rule why @afragen's GitHub Updater plugin isn't allowed.

**Comment by jdevalk** [comment]
I've emailed the .org plugins team the following, subject "Policy question":

> Hi team,
>
> I've got a policy question. We're thinking of building the following thing into both our WordPress SEO > plugin and our Google Analytics plugin:
> 
> - a list of our premium extensions that links to yoast.com to buy them (well, no need to build, we've got that, but it's part of the "flow")
> - an input field where people could put in their "serial" for the product they bought on yoast.com
> - code in WordPress SEO that would then fetch the plugin files from yoast.com and install it (as a separate plugin) in their WordPress install.
> 
> The reason for building this is that giving people a zip file for a premium plugin to install leads to a lot of support work and leads to people having old versions of plugins etc. This would take a lot of the pain away, but of course we  want to stay inside .org rules.
> 
> Would love to hear what, if any, your concerns around this would be!
> 
> Best,
> Joost

**Comment by jdevalk** [comment]
Ugh, answer:

> We don't allow code in the plugin directory to install plugins from elsewhere. This is a hard and fast rule... if we start allowing that sort of thing, then authors will use it to install their code from elsewhere just to get around our rules. They've tried before, we actively monitor the directory to prevent just that sort of thing.
>
> What I would suggest instead is still to have them download a plugin from your site and install it as a ZIP, but you only really need one plugin for them to install manually to handle that. That plugin can then do things like installing those other plugins, keeping them up-to-date, etc. All your serial checking, installation, and update logic can be in that single "Yoast Premium Installer" plugin.

As we already do the latter, do we have suggestions?

**Comment by jdevalk** [comment]
We could give the zip file URL in the backend (not download it but at least they'd always get the latest version etc.)

**Comment by afragen** [comment]
@jdevalk why not create an installer similar to GitHub Updater's remote install feature? You would, in theory, only have to pass the plugin repo directory or .zip path to the user and it would used core install routines. Pardon me if this exists or if that's how it currently functions, I have looked in depth at the code.

A remote install from your own server has the advantages of you creating your own API endpoints. Then the same feature coukd be used to install any dependent plugin and only from your server with appropriate query args attached to the endpoint.

**Comment by jdevalk** [comment]
@afragen you're missing the context of the repository you're in :) we have all this already :) I want to make the connection free -> premium easier.

**Comment by afragen** [comment]
Sorry, I'm not understanding the problem. I didn't mean to confuse the issue.

**Comment by coenjacobs** [comment]
> @afragen you're missing the context of the repository you're in :) we have all this already :) I want to make the connection free -> premium easier.

@jdevalk I think @afragen is not missing the point. He's talking about having a separate plugin that covers the functionality as discussed here, aka "an installer". The issue clearly is installing remote code, which is not allowed, so having a separate plugin "Yoast Helper" would be a solution. It's not going to help make it easier for users to install other plugins, so not really solving the issue at hand here.

BUT, WooThemes does this for their Updater plugin as well, they could add functionality as described in this thread (installing remote code, basically). Maybe this is not a viable solution to this problem for you now, but I didn't want to leave it unnamed.
